solr/core/src/java/org/apache/solr/util/circuitbreaker/AveragingMetricProvider.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.util.circuitbreaker;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import java.io.Closeable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.common.util.ExecutorUtil;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.apache.solr.logging.CircularList;
+
+/** Averages the metric value over a period of time */
+public class AveragingMetricProvider implements Closeable {
+  private final CircularList<Double> samplesRingBuffer;
+  private ScheduledExecutorService executor;
+  private final AtomicDouble currentAverageValue = new AtomicDouble(-1);
+
+  /**
+   * Creates an instance with an executor that runs every sampleInterval seconds and averages over
+   * numSamples samples.
+   *
+   * @param metricProvider metric provider that will provide a value
+   * @param numSamples number of samples to calculate average for
+   * @param sampleInterval interval between each sample
+   */
+  public AveragingMetricProvider(
```

Comment:
Perhaps this could be replaced with one of the already existing implementations for EWMAs in the Dropwizard metrics (or metrics that use them, such as Meter).

Reply:
I had a look at Meter, but it seems very event/count oriented, not for smoothing out gauge values? Or can you show examples of such use somewhere?

Reply:
Hmm, not sure what you mean by event/count oriented... it does the counting, true, but it also keeps track of `ExponentialMovingAverages`, which in turn uses a bunch of EWMAs. "Rate" in its terminology is exactly the moving average of values that are recorded into it by `EWMA.update(n)` method (or `Meter.mark(n)`)

Reply:
Had a look, but the `getRate(TimeUnit rateUnit)` implies events/second, so this API feels like the wrong tool. We're not measuring events per second, but average of a gauge over time. I'm sure it could be shoehorned in, but then I don't see a big gain compared to the simple ring buffer?